File: Fortan 77/extract.f
       program extract
c
c Load in the turnsheet instructions
c
       character*80 line
       character*80 info(20)
       character*80 party(20)
       integer a,start
       integer count
       integer value
       integer takesystem
       integer retreat
       integer system
       integer van(5),res1(5)

c
c       print '($,2a)',char(22),char(12) 
c
       a= 98
       print*,char(a)

       print*,ichar('b')
c
       open(unit=20,file='orders',status='old')
       
10     continue
       read(unit=20,fmt='(a80)') line
       
       if (line(1:6).eq.'combat') then
            print*,' '
            print*,' '
            print*,line(1:79)
c
c okay want to extract all the information I can
c
             count=0
             i=0
30           continue
                 i=i+1
                 if (line(i:i).gt.' ') then
                    count=count+1
                    start=i
                    value=ichar(line(i:i))-48
40                  continue 
                    i=i+1
                    if (line(i:i).eq.' ') then
                          info(count)=line(start:i-1)

                          if (count.eq.2) then
                             system=value
                          endif
                  
                          if (count.eq.5) then
                             van(1)=value
                          endif
                          if (count.eq.6) then
                             van(2)=value
                          endif
                          if (count.eq.7) then
                             van(3)=value
                          endif
                          if (count.eq.8) then
                             van(4)=value
                          endif
                          if (count.eq.9) then
                             van(5)=value
                          endif

                          if (count.eq.10) then
                             res1(1)=value
                          endif
                          if (count.eq.11) then
                             res1(2)=value
                          endif
                          if (count.eq.12) then
                             res1(3)=value
                          endif
                          if (count.eq.13) then
                             res1(4)=value
                          endif
                          if (count.eq.14) then
                             res1(5)=value 
                          endif

                          if (count.eq.15) then
                             retreat=value
                          endif
                          if (count.eq.16) then
                             takesystem=value 
                          endif


                          goto 30
                    else
                        value=value*10+ichar(line(i:i))-48
                    endif
                    goto 40
                 endif
                 if (i.ne.80) goto 30
c                 print*,count
                  print*,' ' 
                  print*,'Fighting at system = ',system
                  print*,'The vanguard is : ',
     &                   van(1),van(2),van(3),van(4),van(5)
                  print*,'First reserve is: ',
     &                   res1(1),res1(2),res1(3),res1(4),res1(5)
                  print*,'Retreat factor = ',retreat
                  print*,'Take system    = ',takesystem
c
c okay lets work out who is on the attack list
c
        numgroups=0

        i=0
        start=1
100     continue
        i=i+1
        if (info(3)(i:i).eq.'0') goto 110
        if (info(3)(i:i).eq.',') then           
             numgroups=numgroups+1
             party(numgroups)=info(3)(start:i-1)
             start=i+1
        endif
        goto 100
110     continue

        print*,'Number of groups = ',numgroups
        do 120 k=1,numgroups
c              print*,k,'  ',party(k)
               print*,' '
              i=0
              start=1
200           continue
              i=i+1
              if (party(k)(i:i).eq.' ') then
                  houses=houses+1
                  call atoi( party(k)(start:i-1),value)
                  print*,value
                  goto 210
              endif
                                   
              if (party(k)(i:i).eq.'=') then           
                  houses=houses+1
                  call atoi( party(k)(start:i-1),value)
                  print*,value
                  start=i+1
              endif
              goto 200
210           continue
120     continue

         
  
        endif

c
c If not at the end of the file then continue
c
       if (line(1:1).ne.'.') goto 10

20     continue
       
       stop
       end



c
c subroutine to calcuate the numerical value of a ansi character
c
       subroutine atoi(string,value)
c
       character*80 string
       integer value
       integer i
c
       i = 0            
       value = 0
10     continue
       i=i+1
       if (string(i:i).eq.' ') goto 20
       value = value*10 + ichar(string(i:i))-48

20     continue
       return
       end

File: Python/extract.py
# extract
# Copyright (C) 2026
# e334c174-bd92-4bd0-af6c-a3e0594bfc08
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https:#www.gnu.org/licenses/>.
# These Terms shall be governed and construed in accordance with the laws of 
# England and Wales, without regard to its conflict of law provisions.

from atoi import atoi


def extract():
    info = [""] * 20
    party = [""] * 20

    van = [0] * 5
    res1 = [0] * 5

    takesystem = 0
    retreat = 0
    system = 0

    print(chr(98))          # Fortran: a = 98 ; print*,char(a)
    print(ord('b'))         # Fortran: print*,ichar('b')

    # Read file
    with open("orders", "r") as f:
        for line in f:
            line = line.rstrip("\n")
            if line.startswith("combat"):
                print()
                print()
                print(line[:79])

                # -----------------------------------------
                # Extract tokens from the combat line
                # -----------------------------------------
                count = 0
                i = 0
                length = len(line)

                while i < length:
                    if line[i] != ' ':
                        count += 1
                        start = i
                        value = ord(line[i]) - 48

                        i += 1
                        while i < length and line[i] != ' ':
                            value = value * 10 + (ord(line[i]) - 48)
                            i += 1

                        token = line[start:i]
                        info[count - 1] = token

                        # Map values to variables
                        if count == 2:
                            system = value

                        if 5 <= count <= 9:
                            van[count - 5] = value

                        if 10 <= count <= 14:
                            res1[count - 10] = value

                        if count == 15:
                            retreat = value

                        if count == 16:
                            takesystem = value
                    else:
                        i += 1

                # -----------------------------------------
                # Print extracted data
                # -----------------------------------------
                print()
                print("Fighting at system =", system)
                print("The vanguard is :", van)
                print("First reserve is:", res1)
                print("Retreat factor =", retreat)
                print("Take system    =", takesystem)

                # -----------------------------------------
                # Parse attack groups from info(3)
                # -----------------------------------------
                groups = info[2]
                numgroups = 0
                start = 0
                party_list = []

                for i, ch in enumerate(groups):
                    if ch == ',':
                        numgroups += 1
                        party_list.append(groups[start:i])
                        start = i + 1
                    if ch == '0':
                        break

                print("Number of groups =", numgroups)

                # -----------------------------------------
                # Parse each party group
                # -----------------------------------------
                for group in party_list:
                    print()
                    start = 0
                    for i, ch in enumerate(group + " "):
                        if ch in [' ', '=']:
                            substring = group[start:i]
                            if substring:
                                value = atoi(substring)
                                print(value)
                            start = i + 1

            # End of file condition
            if line.startswith('.'):
                break


# Run the program
if __name__ == "__main__":
    extract()